canonicalize.go

```diff
@@ -84,8 +84,76 @@ func _canonicalHack(el *etree.Element, seenSoFar map[string]bool) *etree.Element
 // any other other shortcomings until I can get the fixes upstream.
 // NOTE(phoebe): Looks like etree also doesn't remove attributes that are
 // duplicate namespaces. They should be removed if a parent has a matching one
+// NOTE(astuart): namespaces must also be moved to the first element at which they're used
 func canonicalHack(el *etree.Element) *etree.Element {
-	attrMap := make(map[string]bool)
+	attrMap := make(map[string]struct{})
 	return _canonicalHack(el, attrMap)
+}
+
+func getNsDecl(a etree.Attr) string {
+	return fmt.Sprintf("xmlns:%s", a.Key)
+}
+
+type c14nSpace struct {
+	a    etree.Attr
+	used bool
+}
+
+const nsSpace = "xmlns"
+
+func _excCanonicalPrep(el *etree.Element, _alreadyDeclared map[string]c14nSpace) *etree.Element {
+	//Copy alreadyDeclared map
+	alreadyDeclared := make(map[string]c14nSpace, len(_alreadyDeclared))
+	for k := range _alreadyDeclared {
+		alreadyDeclared[k] = _alreadyDeclared[k]
+	}
+
+	usedHere := make(map[string]struct{})
+
+	if el.Space != "" {
+		usedHere[el.Space] = struct{}{}
+	}
+
+	toRemove := make([]string, 0, 0)
+
+	for _, a := range el.Attr {
+		switch a.Space {
+		case nsSpace:
+			toRemove = append(toRemove, a.Space+":"+a.Key)
+			if _, ok := alreadyDeclared[a.Key]; !ok {
+				alreadyDeclared[a.Key] = c14nSpace{a: a, used: false}
+			}
+		default:
+			if a.Space != "" {
+				usedHere[a.Space] = struct{}{}
+			}
+		}
+	}
+	for _, attrK := range toRemove {
+		el.RemoveAttr(attrK)
+	}
+
+	for k := range usedHere {
+		spc := alreadyDeclared[k]
+		//If previously unused, mark as used
+		if !spc.used {
+			el.Attr = append(el.Attr, spc.a)
```

Comment:
Oh, and no it shouldn't risk duplicating attributes since it's only tracking the namespace as the map key. Later on we're simply iterating over namespaces and re-adding them if they've been used at this element but not yet declared.